File: sol/120/120.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        memo = [0] * n
        memo[0] = triangle[0][0]
        for i in range(1, n):
            memo[i] = triangle[i][i] + memo[i - 1]
            for j in range(i - 1, 0, -1):
                memo[j] = triangle[i][j] + min(memo[j], memo[j - 1])
            memo[0] += triangle[i][0]
        return min(memo)

File: sol/120/120.cpp
struct Solution {
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size();
        vector<int> memo(n);
        memo[0] = triangle[0][0];
        for (int i = 1; i < n; ++i){
            memo[i] = triangle[i][i] + memo[i - 1];
            for (int j = i - 1; j > 0; --j){
                memo[j] = triangle[i][j] + min(memo[j], memo[j - 1]);
            }
            memo[0] += triangle[i][0];
        }
        return ranges::min(memo);
    }
};
